### ISSUE TITLE
Permission error when on Android behind SNI proxy

### ISSUE THREAD

**Issue by anderspitman** [caddyserver/certmagic#111]
## What is your question?

I'm working on a [client application](https://github.com/boringproxy/boringproxy/blob/35874f7df01ec3129002ecda2d7074aa149f7f3e/client.go) that runs behind what is essentially (for the purposes of my current issue) an SNI proxy. The SNI proxy binds to port 443, and tunnels to the client over SSH (think ngrok). There's a mode where TLS can be terminated on the client, so the server just passes the requests through, enabling e2ee. This is working great for linux-amd64, but android-arm64 is failing with:

```
solving challenges: presenting for challenge: presenting with embedded solver: could not start listener for challenge server at :443: listen tcp :443: bind: permission denied
```

I have the HTTP challenge disabled.

Any guesses? I haven't tried digging in to the certmagic code yet, but I'm surprised it's apparently trying to bind port 443 for the challenge, since I have my own TLS net.Listener which is providing the net.Conn's.

## What have you already tried?

* Linux AMD64 works, and I verified the executable doesn't have permission to bind 443 so if that's really what's happening it should be failing there too, but it's not.
* Android ARM64 works when not doing passthrough, ie it's probably not some other random issue like DNS.
* I set `Default.Storage` to a writable location and verified certmagic is able to write there on Android.
* I'm properly setting the ACME email.

## Bonus: What do you use this package for, and does it help you?

It's the cert backbone for [boringproxy](https://boringproxy.io/) and I would be in a world of hurt without it!

**Comment by mholt** [comment]
CertMagic attempts to bind a listener to solve the challenge if it cannot detect that a listener is already bound:

https://github.com/caddyserver/certmagic/blob/0f8a9f688760e28d269e64777cf9d9684c32ef3a/solvers.go#L541-L596

The full logs might give you some hint though, I doubt that was the only output.

**Comment by anderspitman** [comment]
Excellent, that looks like the ticket. The issue was being shadowed on linux-amd64 because I was running the server and client on the same machine, so when the client went to bind 443 it saw that the server already had it bound and returned without error. I just confirmed that running them on separate machines produces the same behavior as android-arm64.

The next obvious question for me is: how necessary is this check? Would it be possible for certmagic to return without error even if 443 isn't bound, perhaps issuing a warning in this case? Or failing that, could an option be added to instruct it to specifically ignore this?

Unless I'm mistaken, as it currently stands, programs that use certmagic won't be able to work behind services like ngrok unless the process happens to have permission to bind 443.

**Comment by mholt** [comment]
Great, glad that helped!

> The next obvious question for me is: how necessary is this check? 

Required.

> Unless I'm mistaken, as it currently stands, programs that use certmagic won't be able to work behind services like ngrok unless the process happens to have permission to bind 443.

Either set HTTPSPort or AltTLSALPNPort to something else:

https://github.com/caddyserver/certmagic/blob/0f8a9f688760e28d269e64777cf9d9684c32ef3a/certmagic.go#L463-L476

https://github.com/caddyserver/certmagic/blob/0f8a9f688760e28d269e64777cf9d9684c32ef3a/acmemanager.go#L64-L69

**Comment by anderspitman** [comment]
That did the trick, thanks! You've just enabled e2ee in boringproxy.

PS - I tried to sponsor you but my GH account is locked from purchases for some reason. Trying to resolve that now, but do you have another method for receiving payments? Even if I get it working I'll probably only do a single month at $25. Can't really justify that much ongoing until I'm using certmagic in production for something creating revenue.

**Comment by mholt** [comment]
Awesome, that's great!

And thank you, I appreciate it! A one-time contribution could be sent through https://matt.life/pay

PS. Would a layer 4 proxy be of any use to you? (raw TCP / UDP bytes?) https://github.com/mholt/caddy-l4

**Comment by anderspitman** [comment]
I've looked at conncept in the past when you recommended it. Currently I don't need the flexibility it offers. So far I've been able to walk the line between putting too much effort into reinventing the wheel vs taking on a more generic dependency. I may change my mind in the future, but so far it's been a valuable learning experience implementing a proxy mostly from scratch. Or at least as much from scratch as I can claim with go's awesome stdlib. But that might change in the future at which point I'll likely either use conncept or a Caddy plugin.
